### PR TITLE
nixos/prosody: add authentication option (fixes #53134)

### LINKED ISSUE #53134: prosody stores passwords in plain text
## Issue description

The prosody module does not have an option to set the `authentication` config. Prosody stores passwords in plain text by default. 
An option for the `authentication` config should be added with default value `"internal_hashed"`.

To work around this, I added
```
extraConfig = ''
  authentication = "internal_hashed"
'';
```
to my config

### Steps to reproduce

Set up prosody, register, check the user file in the data path. It will contain the password in plain text.

### PR DESCRIPTION
**Pull request by ghost**
Replaces #53308

Passwords should not be stored in plain text by default. On existing
installations the next time a users user accounts will automatically
be upgraded from plain to hashed one-by-one as they log in.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/networking/prosody.nix [hunk 1]
```diff
@@ -422,6 +422,13 @@ in
         description = "List of administrators of the current host";
       };
 
+      authentication = mkOption {
+        type = types.str;
+        default = "internal_hashed";
+        example = "internal_plain";
```
Comment (Infinisil) [inline nixos/modules/services/networking/prosody.nix]: Is `internal_hashed` the default if not set?

And if there are few enough options, a `types.enum [ "internal_hashed" ... ]` would work well here.

File: nixos/modules/services/networking/prosody.nix [hunk 2]
```diff
@@ -422,6 +422,13 @@ in
...
+        description = "Authentication mechanism used for logins";
```
Comment (Infinisil) [inline nixos/modules/services/networking/prosody.nix]: Needs a `.` at the end (by convention :))

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/prosody.nix (hunk 1 above)
Comment (petabyteboy) [reply nixos/modules/services/networking/prosody.nix]: Prosody uses internal_plain as default value if no value is present in the config. Is there a convention to use the same default values the application provides?

File: nixos/modules/services/networking/prosody.nix (hunk 2 above)
Comment (petabyteboy) [reply nixos/modules/services/networking/prosody.nix]: Fixed :)

File: nixos/modules/services/networking/prosody.nix (hunk 1 above)
Comment (Infinisil) [reply nixos/modules/services/networking/prosody.nix]: Not necessarily, but I'm worried about backwards compatibility

Comment (petabyteboy) [reply nixos/modules/services/networking/prosody.nix]: As described in the commit message: Old installations will be changed to internal_hashed, if nothing else was specified in extraConfig. As described [here](https://prosody.im/doc/modules/mod_auth_internal_hashed#activating): "When you first activate hashed storage, user accounts will automatically be upgraded from plain to hashed one-by-one as they log in. "

Comment (Infinisil) [reply nixos/modules/services/networking/prosody.nix]: Ahh sorry, I was confused by it at first, thanks

**Review by infinisil — APPROVED** [review]

**Pull request merged**